filetracker/cachecleaner.py

```diff
@@ -0,0 +1,248 @@
+    """Tool for periodically cleaning cache of the file tracker.
+       Designed to work as a daemon.
+
+       Cache cleaner is run by calling method :func:`run`. It supports
+       multiple instances of :class:`Client`. Configuration is passed
+       as constructors parameters:
+
+       :param int cache_size_limit: soft limit for sum of logical files size
+       :param iterable glob_cache_dirs: list of paths to
+           :class:`filetracker.Client` cache directories as glob expressions
+       :param datetime.timedelta scan_interval: interval specifying how often
+           scan the disk and optionally clean cache
+       :param float percent_cleaning_level: how many percent of
+           ``cache_size_limit`` of newest cache files do not delete during \
+            cleaning cache
+
+       Cache cleaner runs the following algorithm:
+
+       1. Ask each client (specified in constructor by cache directory)
+          to list all stored files. This is *file index*.
+       2. Analyze file index - check whether cache cleaner should clean
+          the cache and what files exactly.
+       3. Clean cache if necessary.
+       4. Wait time specified in constructor.
+       5. Go to step 1.
+
+       Files are being deleting from the oldest to newer ones taking
+       into account modification time. If files have the same modification
+       time, then file with greater size is being deleted before the second
+       one.
+    """
+
+    FileIndexEntry = collections.namedtuple('FileIndexEntry',
+                                            ['file_info', 'client'])
+    """Entry for file index.
+
+        Associates :class:`DataStore.FileInfoEntry` with
+        :class:`filetracker.Client` which owns given file.
+
+        Fields:
+
+        * ``file_info`` instance of :class:`DataStore.FileInfoEntry`
+        * ``client`` instance of :class:`filetracker.Client` which owns given \
+            file
+    """
+
+    def __init__(self, cache_size_limit, glob_cache_dirs,
+                 scan_interval=datetime.timedelta(minutes=10),
+                 percent_cleaning_level=50.0):
+        assert glob_cache_dirs
+        self.glob_cache_dirs = glob_cache_dirs
+        self.clients = []
+        self.file_index = []
+        self.scan_interval = scan_interval
+        self.cache_size_limit = cache_size_limit
+        self.cleaning_level = cache_size_limit * percent_cleaning_level / 100.
+
+    def run(self):
+        """Starts cleaning cache in infinite loop.
+        """
+        logger.info("Starting daemon.")
+        while True:
+            self._scan_disk()
+            do_cleaning, delete_from_index = self._analyze_file_index()
+            if do_cleaning:
+                self._clean_cache(delete_from_index)
+            sleeping_until_time = datetime.datetime.now() + self.scan_interval
+            logger.info("Sleeping until %s.", sleeping_until_time)
+            time.sleep(self.scan_interval.total_seconds())
```

Comment:
Przydałoby się tutaj obsługiwać wyjątki, albo przynajmniej jakoś zapobiec takiej sytuacji:
1. Powiedzmy, że wyjątek wylatuje z self._clean_cache za każdym uruchomieniem.
2. Tak jak jest teraz, to supervisor będzie restartował cache-cleanera, który w kółko będzie skanował i wysypywał się, obciążając serwer i nie robiąc nic pożytecznego.